File: golang/internal/agents/simple_llm/tools.go
package simple_llm

import (
	"encoding/json"
	"fmt"
	"slices"
	"strings"

	openai "github.com/sashabaranov/go-openai"

	"ai-trading-agents/internal/mcp"
)

// MCPToolset manages multiple MCP clients and their tool definitions.
// It discovers tools from each MCP server, converts them to OpenAI format,
// and routes tool calls to the correct client based on tool name prefix.
type MCPToolset struct {
	clients map[string]*mcp.Client // keyed by name: "trading", "memory"
	tools   []mcp.Tool             // all discovered tools across all clients
	// routes maps tool name prefix to client name.
	// e.g., "trade/" -> "trading", "market/" -> "trading", "memory/" -> "memory"
	routes map[string]string
	// openaiToMCP maps sanitized OpenAI function names back to original MCP tool names.
	// OpenAI requires ^[a-zA-Z0-9_-]+$ so slashes are replaced with double underscores.
	openaiToMCP map[string]string
}

// NewMCPToolset creates a new toolset with the given MCP clients.
// Routes map tool name prefixes to client names (e.g., "trade/" -> "trading").
func NewMCPToolset(clients map[string]*mcp.Client, routes map[string]string) *MCPToolset {
	return &MCPToolset{
		clients: clients,
		routes:  routes,
	}
}

// DiscoverTools calls tools/list on each MCP client and collects all tool definitions.
// Clients are iterated in sorted order to produce a deterministic tool list,
// which ensures consistent JSON serialization for LLM prompt caching.
func (ts *MCPToolset) DiscoverTools() error {
	ts.tools = nil

	names := make([]string, 0, len(ts.clients))
	for name := range ts.clients {
		names = append(names, name)
	}
	slices.Sort(names)

	for _, name := range names {
		tools, err := ts.clients[name].ListTools()
		if err != nil {
			return fmt.Errorf("list tools from %s: %w", name, err)
		}
		ts.tools = append(ts.tools, tools...)
	}
	return nil
}

// Tools returns the discovered MCP tool definitions.
func (ts *MCPToolset) Tools() []mcp.Tool {
	return ts.tools
}

// toOpenAIName converts an MCP tool name to a valid OpenAI function name.
// OpenAI requires function names to match ^[a-zA-Z0-9_-]+$ so slashes are replaced.
func toOpenAIName(mcpName string) string {
	return strings.ReplaceAll(mcpName, "/", "__")
}

// ToOpenAITools converts all discovered MCP tools to OpenAI function-calling format.
// Tool names are sanitized (slashes replaced with double underscores) because
// OpenAI only allows alphanumeric characters, underscores, and hyphens.
func (ts *MCPToolset) ToOpenAITools() []openai.Tool {
	ts.openaiToMCP = make(map[string]string, len(ts.tools))
	result := make([]openai.Tool, 0, len(ts.tools))
	for _, t := range ts.tools {
		params := t.InputSchema
		if params == nil {
			params = map[string]any{"type": "object", "properties": map[string]any{}}
		}

		openaiName := toOpenAIName(t.Name)
		ts.openaiToMCP[openaiName] = t.Name

		result = append(result, openai.Tool{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name:        openaiName,
				Description: t.Description,
				Parameters:  params,
			},
		})
	}
	return result
}

// CallTool routes a tool call to the correct MCP client based on the tool name prefix
// and returns the result. The name can be either an OpenAI-sanitized name (with __)
// or the original MCP name (with /). It is resolved back to the MCP name before routing.
func (ts *MCPToolset) CallTool(name string, args map[string]any) (*mcp.ToolResult, error) {
	mcpName := ts.toMCPName(name)

	clientName, err := ts.resolveClient(mcpName)
	if err != nil {
		return nil, err
	}

	client, ok := ts.clients[clientName]
	if !ok {
		return nil, fmt.Errorf("client %q not found for tool %q", clientName, mcpName)
	}

	return client.CallTool(mcpName, args)
}

// CallToolJSON parses a JSON arguments string and routes the tool call.
func (ts *MCPToolset) CallToolJSON(name string, argsJSON string) (*mcp.ToolResult, error) {
	var args map[string]any
	if argsJSON != "" {
		if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
			return nil, fmt.Errorf("parse tool arguments: %w", err)
		}
	}
	return ts.CallTool(name, args)
}

// toMCPName converts an OpenAI function name back to the original MCP tool name.
// Falls back to the input if no mapping is found.
func (ts *MCPToolset) toMCPName(openaiName string) string {
	if mcpName, ok := ts.openaiToMCP[openaiName]; ok {
		return mcpName
	}
	return openaiName
}

// Client returns the MCP client registered under the given name, or nil if not found.
func (ts *MCPToolset) Client(name string) *mcp.Client {
	return ts.clients[name]
}

// resolveClient finds the client name for a given tool name using prefix matching.
func (ts *MCPToolset) resolveClient(toolName string) (string, error) {
	for prefix, clientName := range ts.routes {
		if strings.HasPrefix(toolName, prefix) {
			return clientName, nil
		}
	}
	return "", fmt.Errorf("no MCP client registered for tool %q", toolName)
}

File: rust/src/agents/simple_llm/tools.rs
use std::collections::HashMap;

use anyhow::bail;

use crate::mcp::{McpClient, Tool, ToolResult};

/// Manages multiple MCP clients and routes tool calls by prefix.
/// Mirrors Go's MCPToolset.
pub struct McpToolset {
    clients: HashMap<String, McpClient>,
    tools: Vec<Tool>,
    /// Tool name prefix -> client name. e.g. "trade/" -> "trading"
    routes: HashMap<String, String>,
    /// OpenAI-sanitized name -> original MCP name. e.g. "trade__submit_intent" -> "trade/submit_intent"
    openai_to_mcp: HashMap<String, String>,
}

impl McpToolset {
    pub fn new(clients: HashMap<String, McpClient>, routes: HashMap<String, String>) -> Self {
        Self {
            clients,
            tools: Vec::new(),
            routes,
            openai_to_mcp: HashMap::new(),
        }
    }

    /// Call list_tools on each MCP client and collect all tool definitions.
    pub async fn discover_tools(&mut self) -> anyhow::Result<()> {
        self.tools.clear();
        for (name, client) in &self.clients {
            let tools = client
                .list_tools()
                .await
                .map_err(|e| anyhow::anyhow!("list tools from {name}: {e}"))?;
            self.tools.extend(tools);
        }
        Ok(())
    }

    /// Get all discovered tools.
    pub fn tools(&self) -> &[Tool] {
        &self.tools
    }

    /// Convert MCP tools to OpenAI function-calling format.
    /// Returns JSON array suitable for the `tools` field in ChatCompletionRequest.
    pub fn to_openai_tools(&mut self) -> Vec<serde_json::Value> {
        self.openai_to_mcp.clear();
        let mut result = Vec::with_capacity(self.tools.len());

        for t in &self.tools {
            let schema = t.input_schema.clone().unwrap_or_else(|| {
                serde_json::json!({"type": "object", "properties": {}})
            });

            let openai_name = to_openai_name(&t.name);
            self.openai_to_mcp
                .insert(openai_name.clone(), t.name.clone());

            result.push(serde_json::json!({
                "type": "function",
                "function": {
                    "name": openai_name,
                    "description": t.description.as_deref().unwrap_or(""),
                    "parameters": schema,
                }
            }));
        }

        result
    }

    /// Route a tool call to the correct MCP client by prefix.
    pub async fn call_tool(
        &self,
        name: &str,
        args: Option<serde_json::Map<String, serde_json::Value>>,
    ) -> anyhow::Result<ToolResult> {
        let mcp_name = self.to_mcp_name(name);
        let client_name = self.resolve_client(&mcp_name)?;

        let client = self
            .clients
            .get(&client_name)
            .ok_or_else(|| anyhow::anyhow!("client {client_name:?} not found for tool {mcp_name:?}"))?;

        client.call_tool(&mcp_name, args).await
    }

    /// Parse JSON args string and route the tool call.
    pub async fn call_tool_json(
        &self,
        name: &str,
        args_json: &str,
    ) -> anyhow::Result<ToolResult> {
        let args = if args_json.is_empty() {
            None
        } else {
            let map: serde_json::Map<String, serde_json::Value> =
                serde_json::from_str(args_json)
                    .map_err(|e| anyhow::anyhow!("parse tool arguments: {e}"))?;
            Some(map)
        };
        self.call_tool(name, args).await
    }

    /// Get a client by name (e.g. "files").
    pub fn client(&self, name: &str) -> Option<&McpClient> {
        self.clients.get(name)
    }

    fn to_mcp_name(&self, openai_name: &str) -> String {
        self.openai_to_mcp
            .get(openai_name)
            .cloned()
            .unwrap_or_else(|| openai_name.to_string())
    }

    fn resolve_client(&self, tool_name: &str) -> anyhow::Result<String> {
        for (prefix, client_name) in &self.routes {
            if tool_name.starts_with(prefix) {
                return Ok(client_name.clone());
            }
        }
        bail!("no MCP client registered for tool {tool_name:?}");
    }
}

/// Convert MCP name to OpenAI-safe name: "/" -> "__"
fn to_openai_name(mcp_name: &str) -> String {
    mcp_name.replace('/', "__")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::mcp::Tool;
    use std::time::Duration;
    use wiremock::matchers::method;
    use wiremock::{Mock, MockServer, ResponseTemplate};

    #[test]
    fn test_to_openai_name() {
        assert_eq!(to_openai_name("trade/submit_intent"), "trade__submit_intent");
        assert_eq!(to_openai_name("memory/read"), "memory__read");
        assert_eq!(to_openai_name("no_slash"), "no_slash");
    }

    #[test]
    fn test_resolve_client() {
        let routes = HashMap::from([
            ("trade/".to_string(), "trading".to_string()),
            ("market/".to_string(), "trading".to_string()),
            ("memory/".to_string(), "memory".to_string()),
        ]);

        let ts = McpToolset::new(HashMap::new(), routes);

        assert_eq!(ts.resolve_client("trade/submit_intent").unwrap(), "trading");
        assert_eq!(ts.resolve_client("market/get_prices").unwrap(), "trading");
        assert_eq!(ts.resolve_client("memory/read").unwrap(), "memory");
        assert!(ts.resolve_client("unknown/tool").is_err());
    }

    #[test]
    fn test_to_openai_tools() {
        let tools = vec![
            Tool {
                name: "trade/submit_intent".to_string(),
                description: Some("Submit a trade".to_string()),
                input_schema: Some(serde_json::json!({
                    "type": "object",
                    "properties": {
                        "market": {"type": "string"}
                    }
                })),
            },
            Tool {
                name: "memory/read".to_string(),
                description: Some("Read memory".to_string()),
                input_schema: None,
            },
        ];

        let mut ts = McpToolset::new(HashMap::new(), HashMap::new());
        ts.tools = tools;

        let openai_tools = ts.to_openai_tools();
        assert_eq!(openai_tools.len(), 2);

        // Check first tool
        let first = &openai_tools[0];
        assert_eq!(first["type"], "function");
        assert_eq!(first["function"]["name"], "trade__submit_intent");
        assert_eq!(first["function"]["description"], "Submit a trade");

        // Check name mapping
        assert_eq!(
            ts.to_mcp_name("trade__submit_intent"),
            "trade/submit_intent"
        );
        assert_eq!(ts.to_mcp_name("memory__read"), "memory/read");

        // Fallback for unknown names
        assert_eq!(ts.to_mcp_name("direct/name"), "direct/name");
    }

    #[tokio::test]
    async fn test_discover_tools() {
        let server = MockServer::start().await;

        Mock::given(method("POST"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "jsonrpc": "2.0",
                "id": 1,
                "result": {
                    "tools": [
                        {"name": "trade/submit_intent", "description": "Submit"},
                        {"name": "trade/get_portfolio", "description": "Portfolio"}
                    ]
                }
            })))
            .mount(&server)
            .await;

        let client = McpClient::new(&server.uri(), "", Duration::from_secs(5));
        let clients = HashMap::from([("trading".to_string(), client)]);
        let routes = HashMap::from([("trade/".to_string(), "trading".to_string())]);

        let mut ts = McpToolset::new(clients, routes);
        ts.discover_tools().await.unwrap();

        assert_eq!(ts.tools().len(), 2);
        assert_eq!(ts.tools()[0].name, "trade/submit_intent");
    }

    #[tokio::test]
    async fn test_call_tool_routing() {
        let trading_server = MockServer::start().await;
        let memory_server = MockServer::start().await;

        Mock::given(method("POST"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "jsonrpc": "2.0",
                "id": 1,
                "result": {
                    "content": [{"type": "text", "text": "traded"}],
                    "isError": false
                }
            })))
            .mount(&trading_server)
            .await;

        Mock::given(method("POST"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({
                "jsonrpc": "2.0",
                "id": 1,
                "result": {
                    "content": [{"type": "text", "text": "remembered"}],
                    "isError": false
                }
            })))
            .mount(&memory_server)
            .await;

        let clients = HashMap::from([
            ("trading".to_string(), McpClient::new(&trading_server.uri(), "", Duration::from_secs(5))),
            ("memory".to_string(), McpClient::new(&memory_server.uri(), "", Duration::from_secs(5))),
        ]);
        let routes = HashMap::from([
            ("trade/".to_string(), "trading".to_string()),
            ("memory/".to_string(), "memory".to_string()),
        ]);

        let mut ts = McpToolset::new(clients, routes);
        // Set up openai_to_mcp mapping
        ts.openai_to_mcp.insert("trade__submit_intent".to_string(), "trade/submit_intent".to_string());
        ts.openai_to_mcp.insert("memory__read".to_string(), "memory/read".to_string());

        let result = ts.call_tool("trade__submit_intent", None).await.unwrap();
        assert_eq!(result.content[0].text.as_deref(), Some("traded"));

        let result = ts.call_tool("memory__read", None).await.unwrap();
        assert_eq!(result.content[0].text.as_deref(), Some("remembered"));
    }
}
